# MLZoomcamp 2025 -- Homework #1

In [1]:
import pandas as pd
import numpy as np
import seaborn as sb
import matplotlib as mp

## Q1. Pandas version
What version of Pandas did you install?

In [2]:
pd.__version__

'2.3.2'

## Get data

For this homework, we'll use the Car Fuel Efficiency dataset. You can download it using the wget command.

In [3]:
# !wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv

--2025-10-02 16:35:26--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 2606:50c0:8003::154, 2606:50c0:8000::154, 2606:50c0:8001::154, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|2606:50c0:8003::154|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 874188 (854K) [text/plain]
Saving to: ‘car_fuel_efficiency.csv.1’

car_fuel_efficiency 100%[===================>] 853.70K  --.-KB/s    in 0.09s   

2025-10-02 16:35:27 (9.00 MB/s) - ‘car_fuel_efficiency.csv.1’ saved [874188/874188]



In [48]:
df = pd.read_csv("car_fuel_efficiency.csv")

## Q2. Records count
How many records are in the dataset?

In [49]:
df

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                     170            3.0       159.0     3413.433759   
1                     130            5.0        97.0     3149.664934   
2                     170            NaN        78.0     3079.038997   
3                     220            4.0         NaN     2542.392402   
4                     210            1.0       140.0     3460.870990   
...                   ...            ...         ...             ...   
9699                  140            5.0       164.0     2981.107371   
9700                  180            NaN       154.0     2439.525729   
9701                  220            2.0       138.0     2583.471318   
9702                  230            4.0       177.0     2905.527390   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year  origin fuel_type         drivetrain  \
0             17.7        2003  Europe  Gasoline    All-wheel drive   
1             17.8        2007     USA  Gasoline  Front-wheel drive   
2             15.1        2018  Europe  Gasoline  Front-wheel drive   
3             20.2        2009     USA    Diesel    All-wheel drive   
4             14.4        2009  Europe  Gasoline    All-wheel drive   
...            ...         ...     ...       ...                ...   
9699          17.3        2013  Europe    Diesel  Front-wheel drive   
9700          15.0        2004     USA  Gasoline    All-wheel drive   
9701          15.1        2008     USA    Diesel    All-wheel drive   
9702          19.4        2011     USA    Diesel  Front-wheel drive   
9703          14.7        2005    Asia    Diesel    All-wheel drive   

      num_doors  fuel_efficiency_mpg  
0           0.0            13.231729  
1           0.0            13.688217  
2           0.0            14.246341  
3           2.0            16.912736  
4           2.0            12.488369  
...         ...                  ...  
9699        NaN            15.101802  
9700        0.0            17.962326  
9701       -1.0            17.186587  
9702        1.0            15.331551  
9703       -1.0            14.884467  

[9704 rows x 11 columns]

There are 9704 records in the dataset.

## Q3. Fuel types
How many fuel types are presented in the dataset?

In [50]:
df.fuel_type.unique()

array(['Gasoline', 'Diesel'], dtype=object)

There are two fuel types present.

## Q4. Missing values
How many columns in the dataset have missing values?

In [51]:
count = 0

for column in df:
    if df[column].isnull().sum() > 0:
        print(f"{column} contains {df[column].isnull().sum()} null values")
        count = count + 1
print()
print(f"A total of {count} columns contain missing data.")

num_cylinders contains 482 null values
horsepower contains 708 null values
acceleration contains 930 null values
num_doors contains 502 null values

A total of 4 columns contain missing data.


## Q5. Max fuel efficiency
What's the maximum fuel efficiency of cars from Asia?

In [52]:
df_asia = df.loc[df['origin'] == 'Asia']
df_asia

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
8                     250            1.0       174.0     2714.219310   
12                    320            5.0       145.0     2783.868974   
14                    200            6.0       160.0     3582.687368   
20                    150            3.0       197.0     2231.808142   
21                    160            4.0       133.0     2659.431451   
...                   ...            ...         ...             ...   
9688                  260            4.0         NaN     3948.404625   
9692                  180            3.0       188.0     3680.341381   
9693                  280            2.0       148.0     2545.070139   
9698                  180            1.0       131.0     3107.427820   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year origin fuel_type         drivetrain  num_doors  \
8             10.3        2016   Asia    Diesel  Front-wheel drive       -1.0   
12            15.1        2010   Asia    Diesel    All-wheel drive        1.0   
14            14.9        2007   Asia    Diesel    All-wheel drive        0.0   
20            18.7        2011   Asia  Gasoline  Front-wheel drive        1.0   
21             NaN        2016   Asia  Gasoline  Front-wheel drive       -1.0   
...            ...         ...    ...       ...                ...        ...   
9688          15.5        2018   Asia    Diesel    All-wheel drive       -1.0   
9692          18.0        2016   Asia  Gasoline  Front-wheel drive        1.0   
9693          15.6        2012   Asia    Diesel    All-wheel drive        0.0   
9698          13.2        2005   Asia  Gasoline  Front-wheel drive       -2.0   
9703          14.7        2005   Asia    Diesel    All-wheel drive       -1.0   

      fuel_efficiency_mpg  
8               16.823554  
12              16.175820  
14              11.871091  
20              18.889083  
21              16.077730  
...                   ...  
9688            11.054830  
9692            11.711653  
9693            17.202782  
9698            13.933716  
9703            14.884467  

[3247 rows x 11 columns]

In [53]:
print(f"The maximum fuel efficiency of cars from Asia is {round(df_asia.fuel_efficiency_mpg.max(), 2)}.")

The maximum fuel efficiency of cars from Asia is 23.76.


## Q6. Median value of horsepower
1. Find the median value of the horsepower column in the dataset.
2. Next, calculate the most frequent value of the same horsepower column.
3.  Use the fillna method to fill the missing values in the horsepower column with the most frequent value from the previous step.
4.  Now, calculate the median value of horsepower once again.

Has it changed?

In [54]:
print(f"The median horsepower value in the dataset is {df.horsepower.median()}.")

The median horsepower value in the dataset is 149.0.


In [55]:
print(f"The most common horsepower value in the dataset is {float(df.horsepower.value_counts().index.tolist()[0])}.")

The most common horsepower value in the dataset is 152.0.


In [56]:
df.horsepower.isnull().sum()

np.int64(708)

In [57]:
df.horsepower[3]

np.float64(nan)

In [58]:
df.horsepower = df.horsepower.fillna(value=152.0)

In [59]:
df.horsepower[3]

np.float64(152.0)

In [60]:
print(f"After filling NaN values with the most common horsepower value, the median horsepower value in the dataset is {df.horsepower.median()}.")

After filling NaN values with the most common horsepower value, the median horsepower value in the dataset is 152.0.


Yes--the median horsepower value changed from 149 to 152.

## Q7. Sum of weights
1.  Select all the cars from Asia
2.  Select only columns vehicle_weight and model_year
3.  Select the first 7 values
4.  Get the underlying NumPy array. Let's call it X.
5.  Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6.  Invert XTX.
7.  Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
8.  Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
9.  What's the sum of all the elements of the result?

(Note: You just implemented linear regression. We'll talk about it in the next lesson.)

In [66]:
df_asia_sum_wts = df.loc[df.origin == 'Asia'] 

In [67]:
df_asia_sum_wts = df_asia_sum_wts[['vehicle_weight', 'model_year']]

In [70]:
df_asia_sum_wts[0:7]

vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019

In [73]:
X = df_asia_sum_wts[0:7].to_numpy()

In [90]:
X

array([[2714.21930965, 2016.        ],
       [2783.86897424, 2010.        ],
       [3582.68736772, 2007.        ],
       [2231.8081416 , 2011.        ],
       [2659.43145076, 2016.        ],
       [2844.22753389, 2014.        ],
       [3761.99403819, 2019.        ]])

In [98]:
X.T

array([[2714.21930965, 2783.86897424, 3582.68736772, 2231.8081416 ,
        2659.43145076, 2844.22753389, 3761.99403819],
       [2016.        , 2010.        , 2007.        , 2011.        ,
        2016.        , 2014.        , 2019.        ]])

In [100]:
XTX = X.T.dot(X)
XTX

array([[62248334.33150762, 41431216.5073268 ],
       [41431216.5073268 , 28373339.        ]])

In [101]:
XTX_inv = np.linalg.inv(XTX)

In [102]:
XTX_inv

array([[ 5.71497081e-07, -8.34509443e-07],
       [-8.34509443e-07,  1.25380877e-06]])

In [103]:
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [108]:
w = XTX_inv.dot(X.T) * y

In [110]:
w

array([[-0.14432288, -0.11230828,  0.29810794, -0.36245398, -0.16251372,
        -0.06075771,  0.55811286],
       [ 0.28890053,  0.2560879 , -0.37871378,  0.59305003,  0.30835783,
         0.16679975, -0.72957556]])

In [111]:
w.sum()

np.float64(0.5187709081074023)